### PR TITLE
refactor!: Remove gnu.trove from RowSet API

### PR DESCRIPTION
**Pull request by niloc132**
Removes usages of gnu.trove from the API of rowset related classes. These usages were previously unused, though this is technically an API breaking change.

Partial #188

### CODE REVIEW

File: engine/table/src/main/java/io/deephaven/engine/table/impl/util/ShiftCallback.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+//
+// Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.engine.table.impl.util;
+
+public interface ShiftCallback {
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/util/ShiftCallback.java:6]: I think `ShiftCallback` is unused and should be removed.
We still have `RowSetShiftData.Callback`, which is somewhat used, and should maybe be implemented in place of `ShiftCallback`.
I think whichever survives need a bit of JavaDoc.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/util/RowSetShiftCallback.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+//
+// Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.engine.table.impl.util;
+
+import io.deephaven.engine.rowset.RowSet;
+
+public interface RowSetShiftCallback {
+    void shift(RowSet rowSet, long offset);
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/util/RowSetShiftCallback.java:9]: I think this should just be a (JavaDoc'd) method on `WritableColumnSource`, and we should delete the interface.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/util/ShiftCallback.java (hunk 1 above)
Comment (niloc132) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/util/ShiftCallback.java:6]: Still used it appears:
Implemented in the following places:

- ImmutableConstantFloatSource (io.deephaven.engine.table.impl.sources.immutable)
- ArrayBackedColumnSource (io.deephaven.engine.table.impl.sources)
- NullValueColumnSource (io.deephaven.engine.table.impl.sources)
- SingleValueColumnSource (io.deephaven.engine.table.impl.sources)
- ImmutableConstantByteSource (io.deephaven.engine.table.impl.sources.immutable)
- ImmutableConstantShortSource (io.deephaven.engine.table.impl.sources.immutable)
- ImmutableConstantCharSource (io.deephaven.engine.table.impl.sources.immutable)
- ImmutableConstantDoubleSource (io.deephaven.engine.table.impl.sources.immutable)
- ImmutableConstantIntSource (io.deephaven.engine.table.impl.sources.immutable)
- ImmutableConstantLongSource (io.deephaven.engine.table.impl.sources.immutable)
- NanosBasedTimeArraySource (io.deephaven.engine.table.impl.sources)
- ImmutableConstantObjectSource (io.deephaven.engine.table.impl.sources.immutable)
- ImmutableConstantNanosBasedTimeSource (io.deephaven.engine.table.impl.sources.immutable)


Only called from one place:
NanoBasedTimeArraySource.java
```
    @Override
    public void shift(long start, long end, long offset) {
        nanoSource.shift(start, end, offset);
    }
```

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSetFactory.java [hunk 3]
```diff
@@ -56,21 +54,6 @@ public static WritableRowSet fromKeys(final long rowKey) {
...
 
-    /**
-     * Get a {@link WritableRowSet} containing the specified row keys.
-     * <p>
-     * The provided {@link TLongList} is sorted and then passed to a {@link RowSetBuilderSequential}.
-     *
-     * @param rowKeys A {@link TLongList}. Note that this list is mutated within the method!
-     * @return A new {@link WritableRowSet} containing the values from {@code rowKeys}
-     */
-    public static RowSet fromKeys(@NotNull final TLongList rowKeys) {
```
[the remark this answers is not in the document]
Comment (niloc132) [reply engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSetFactory.java:67]: The RowSetFactory.fromKeys(int...) overload is used in core+, but I see no sign of the TLongList overload.

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSetBuilderSequential.java [hunk 4]
```diff
@@ -17,7 +16,7 @@
 /**
  * Builder interface for {@link RowSet} construction in strict sequential order.
  */
-public interface RowSetBuilderSequential extends TLongProcedure, LongRangeConsumer {
+public interface RowSetBuilderSequential extends LongRangeConsumer {
```
[the remark this answers is not in the document]
Comment (niloc132) [reply engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSetBuilderSequential.java:19]: TLongProcedure.execute is never used in core+.

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSet.java [hunk 5]
```diff
@@ -116,15 +115,6 @@ default WritableRowSet invert(RowSet keys) {
...
     WritableRowSet invert(RowSet keys, long maximumPosition);
 
-    /**
-     * For the given keys RowSet, under the assertion that none of them are present in the current RowSet, return the
-     * tentative insertion points in the current RowSet with the count for each of them
-     *
-     * @param keys the keys to identify insertion locations
-     * @return two TLongArrayLists; [0] contains the positions, [1] contains the counts.
-     */
-    TLongArrayList[] findMissing(RowSet keys);
```
Comment (devinrsmith) [inline engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSet.java:126]: If this is the only place where the API dependency is, please update or remove the dependency in `engine/rowset/build.gradle`

Comment (niloc132) [reply engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSet.java:126]: It looks like we might be able to get away with downgrading that from api to implementation, I'll look into that further.

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSetShiftData.java [hunk 6]
```diff
@@ -224,8 +225,12 @@ public boolean equals(final Object obj) {
      */
     public static final RowSetShiftData EMPTY = new RowSetShiftData();
 
+    /**
+     * Functional interface to pass to {@link RowSetShiftData#apply(RowKeyRangeShiftCallback)} or
+     * {@link RowSetShiftData#unapply(RowKeyRangeShiftCallback)} to get information about each shift recorded.
+     */
     @FunctionalInterface
-    public interface Callback {
+    public interface RowKeyRangeShiftCallback {
```
Comment (rcaudy) [inline engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSetShiftData.java:233]: I thought you were moving this to top level.

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSetShiftCallback.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+ * necessary.
+ */
+public interface RowSetShiftCallback {
+    /**
+     * Signals that the range should be shifted by the provided offset.
+     *
+     * @param rowSet The keys to shift
+     * @param offset The offset to move each key by
+     */
+    void shift(RowSet rowSet, long offset);
```
Comment (rcaudy) [inline engine/rowset/src/main/java/io/deephaven/engine/rowset/RowSetShiftCallback.java:17]:
```suggestion
 * Callback interface for propagating shifts over entire {@link RowSet RowSets}.
 */
public interface RowSetShiftCallback {
    /**
     * Signals that the row keys in {@code rowSet} should be shifted by the provided {@code shiftDelta}.
     *
     * @param rowSet The row keys to shift
     * @param shiftDelta The shift delta to apply to each row key in {@code rowSet}
     */
    void shift(RowSet rowSet, long shiftDelta);
```

**Review by nbauernfeind — APPROVED** [review]

**Review by rcaudy — APPROVED** [review]

**Pull request merged**